lib/spack/spack/test/variant.py

```diff
@@ -0,0 +1,656 @@
+            assert repr(a) == "BoolValuedVariant('foo', {0})".format(repr(v))
+            assert str(a) == '+foo'
+            assert a.value is True
+            assert True in a
+            assert eval(repr(a)) == a
+
+        # Copy - True value
+        b = a.copy()
+        assert repr(a) == repr(b)
+        assert str(a) == str(b)
+        assert b.value is True
+        assert True in b
+        assert a == b
+        assert a is not b
+        assert hash(a) == hash(b)
+        assert eval(repr(b)) == a
+
+        # Basic properties - False value
+        for v in (False, 'False', 'FALSE', 'FaLsE'):
+            a = BoolValuedVariant('foo', v)
+            assert repr(a) == "BoolValuedVariant('foo', {0})".format(repr(v))
+            assert str(a) == '~foo'
+            assert a.value is False
+            assert False in a
+            assert eval(repr(a)) == a
+
+        # Copy - True value
+        b = a.copy()
+        assert repr(a) == repr(b)
+        assert str(a) == str(b)
+        assert b.value is False
+        assert False in b
+        assert a == b
+        assert a is not b
+        assert eval(repr(b)) == a
+
+        # Invalid values
+        for v in ('bar', 'bar,baz'):
+            with pytest.raises(ValueError):
+                BoolValuedVariant('foo', v)
+
+    def test_satisfies(self):
+        a = BoolValuedVariant('foo', True)
+        b = BoolValuedVariant('foo', False)
+        c = BoolValuedVariant('fee', False)
+        d = BoolValuedVariant('foo', 'True')
+
+        assert not a.satisfies(b)
+        assert not a.satisfies(c)
+        assert a.satisfies(d)
+
+        assert not b.satisfies(a)
+        assert not b.satisfies(c)
+        assert not b.satisfies(d)
+
+        assert not c.satisfies(a)
+        assert not c.satisfies(b)
+        assert not c.satisfies(d)
+
+        assert d.satisfies(a)
+        assert not d.satisfies(b)
+        assert not d.satisfies(c)
+
+        # Cannot satisfy the constraint with an object of
+        # another type
+        d_mv = MultiValuedVariant('foo', 'True')
+        assert not d.satisfies(d_mv)
+
+        d_sv = SingleValuedVariant('foo', 'True')
+        assert not d.satisfies(d_sv)
+
+    def test_compatible(self):
+
+        a = BoolValuedVariant('foo', True)
+        b = BoolValuedVariant('fee', True)
+        c = BoolValuedVariant('foo', False)
+        d = BoolValuedVariant('foo', 'True')
+
+        # If the name of two multi-valued variants is the same,
+        # they are compatible
+        assert not a.compatible(b)
+        assert not a.compatible(c)
+        assert a.compatible(d)
+
+        assert not b.compatible(a)
+        assert not b.compatible(c)
+        assert not b.compatible(d)
+
+        assert not c.compatible(a)
+        assert not c.compatible(b)
+        assert not c.compatible(d)
+
+        assert d.compatible(a)
+        assert not d.compatible(b)
+        assert not d.compatible(c)
+
+        # Can't be compatible with other types
```

Comment:
Any chance this could be coercible?  It seems confusing that `pkg foo=true` is not the same as `pkg +foo`.  Or am I jumping the gun here, and this is already handled because both are always parsed as `BooleanValuedVariant`?